src/fabric/src/fabric2_fdb.erl

```diff
@@ -1361,21 +1426,66 @@ get_fold_opts(RangePrefix, Options) ->
         B when is_boolean(B) -> [{snapshot, B}]
     end,
 
-    OutOpts = [{reverse, Reverse}]
-            ++ Limit
+    BaseOpts = [{reverse, Reverse}]
             ++ TargetBytes
             ++ StreamingMode
             ++ Snapshot,
 
-    {StartKey3, EndKey3, Skip, OutOpts}.
+    BaseAcc#fold_acc{
+        start_key = StartKey3,
+        end_key = EndKey3,
+        skip = Skip,
+        limit = Limit,
+        base_opts = BaseOpts,
+        user_fun = UserCallback,
+        user_acc = UserAcc
+    }.
 
 
-fold_range_cb(KV, {skip, 0, Callback, Acc}) ->
-    NewAcc = Callback(KV, Acc),
-    {skip, 0, Callback, NewAcc};
+fold_range_cb({K, V}, #fold_acc{} = Acc) ->
+    #fold_acc{
+        skip = Skip,
+        limit = Limit,
+        user_fun = UserFun,
+        user_acc = UserAcc,
+        base_opts = Opts
+    } = Acc,
+    Acc1 = case Skip =:= 0 of
+        true ->
+            UserAcc1 = UserFun({K, V}, UserAcc),
+            Acc#fold_acc{limit = Limit - 1, user_acc = UserAcc1};
+        false ->
+            Acc#fold_acc{skip = Skip - 1, limit = Limit - 1}
+    end,
+    Acc2 = case fabric2_util:get_value(reverse, Opts, false) of
+        true -> Acc1#fold_acc{end_key = K};
+        false ->Acc1#fold_acc{start_key = K}
+    end,
+    put(?PDICT_FOLD_ACC_STATE, Acc2),
+    Acc2.
 
-fold_range_cb(_KV, {skip, N, Callback, Acc}) when is_integer(N), N > 0 ->
-    {skip, N - 1, Callback, Acc}.
+
+restart_fold(Tx, #fold_acc{} = Acc) ->
+    erase(?PDICT_CHECKED_MD_IS_CURRENT),
+    ok = erlfdb:reset(Tx),
+    case get(?PDICT_FOLD_ACC_STATE) of
+        #fold_acc{db = Db0} = Acc1 ->
+            erase(?PDICT_FOLD_ACC_STATE),
+            Db = check_db_instance(Db0),
+            Acc1#fold_acc{db = Db};
+        undefined ->
+            #fold_acc{
+                db = Db,
+                start_key = Start,
+                end_key = End,
+                limit = Limit,
+                user_fun = UserFun
+            } = Acc,
+            Db = check_db_instance(Acc#fold_acc.db),
+            Args = [Db, Start, End, Limit, UserFun],
+            couch_log:error("fold_range not progressing ~p ~p:~p ~p ~p", Args),
```

Comment:
Seems like this should be conditioned on something? The initial call to fold_range could just be timed badly at the end of a transaction and we'd log this which seems not right.

Also we should probably blow up if we actually do detect some sort of infinite loop here rather than logging that we're in an infinite loop.

Reply:
Good idea. I was thinking we'd set PDICT_FOLD_ACC_STATE to `initialized` in the first call to fold_range then detect in the restart_fold